### DISCUSSION TITLE
How can I pass parameter that would be needed to loop over each data represent by each test case ("it" function)?

### DISCUSSION THREAD

**Discussion by andikabhaskara** [cypress-io/cypress#28558] [Questions and Help]
Hi guys,

so I have been thinking to try to create test case to be more dynamic based on the 1st test case of how much data do I have and I tried to validate it in batches by divide it from the total data that I have. the reason behind this is to reduce the memory consumption and to mitigate the Crash error from Chromium. but, so far I have been facing a problem in my code as follow:

```
describe("Issue fetch data", () => {
  let totalData: number;
  let maxBatches: number;

it("fetch total data", () => {
  cy.fixture("token").then((token) => {
    // logic to fetch total data
    ).then((res) => {
      totalData = res.body.total;
      maxBatches = Math.ceil(totalData / 1000);
      Cypress.env("testdata", maxBatches)
    });
  });
});

it("check if testdata shown", () => {
  cy.log(Cypress.env("testdata")); // it does shown
});

// But when it comes to here, this test case does not shown up.
// only "fetch total data" & "check if testdata shown"

for (let i = 1; i <= Cypress.env("testdata"); i++) {
  it(`get data on batch ${i}`, () => {
    // my code logic
  });
}
```

The problem that I am facing is that, the loop over test case it does not appears, unless I hard coded the max limit loop. I have tried  to follow the instruction in [here](https://glebbahmutov.com/blog/pass-values-between-tests/) so far no luck.

<img width="1445" alt="Screenshot 2023-12-21 at 12 21 02" src="https://github.com/cypress-io/cypress/assets/64412963/b5646621-9a86-4a9e-95d8-cdf33bcf1825">

I am expecting something like this. Note: I hard coded the loop from `for (let i = 1; i <= Cypress.env("testdata"); i++) ` to `for (let i = 1; i <= 2; i++) {`

<img width="977" alt="Screenshot 2023-12-21 at 12 22 03" src="https://github.com/cypress-io/cypress/assets/64412963/09c9819e-87af-474c-af25-e4a1c236d8c5">

Does anyone know alternative to do this? I have also tried to create another param above for loop to be the maximum limit looping, but it still doesn't works. Thank you guys!

**Comment by portyu9** [comment]
Hi @andikabhaskara 

The problem is timing: Mocha registers all of the `it()` blocks while the spec file is being evaluated.

Your first `it()` runs **after** that registration has already happened, so setting a value there cannot cause this later loop to create new tests:

```js
for (let i = 1; i <= maxBatches; i++) {
  it(...)
}
```

If the batch count comes from an API and you want one `it()` per batch, fetch that data in `setupNodeEvents()` **before the spec is loaded**, then expose the non-sensitive result synchronously to the spec.

On current Cypress 16, for example:

```js
// cypress.config.js
const { defineConfig } = require('cypress')

module.exports = defineConfig({
  e2e: {
    async setupNodeEvents(on, config) {
      const response = await fetch('https://example.com/api/data')
      const data = await response.json()

      config.expose = {
        ...config.expose,
        maxBatches: Math.ceil(data.total / 1000),
      }

      return config
    },
  },
})
```

Then generate the tests while the spec is being evaluated:

```js
const maxBatches = Cypress.expose('maxBatches')

describe('data batches', () => {
  for (let i = 1; i <= maxBatches; i++) {
    it(`gets data for batch ${i}`, () => {
      // test batch i
    })
  }
})
```

This is the same lifecycle used by Cypress's official **Dynamic tests from API** recipe.

One version-related note: older examples use `config.env` + `Cypress.env()`. `Cypress.env()` was deprecated in Cypress 15.10 and **removed in Cypress 16**, so for a non-sensitive value such as `maxBatches`, `config.expose` + `Cypress.expose()` is the current API.

The other valid design is to keep a single `it()` and perform the batches inside that test, but if you specifically need separate Mocha test cases, the data must be available before the tests are registered.

Please mark this answer as accepted if it helped, thank you.
